### DISCUSSION TITLE
Noise in Tiled Pyramid TIFF

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#4090] [Help]
Hi all,

we analyze tiff images from the medical field and we are using ImageMagick to create Tiled Pyramid TIFFs for zooming.
```
convert s -define tiff:tile-geometry=512x512 'ptif:o.tif'
```
We have noticed that when you zoom out, i.e. when you look at a smaller layer in pyramidal tiif image, there is a noise in the image.

Pyramid layer without noise:
![WithoutNoise](https://user-images.githubusercontent.com/73162709/129896767-6f8d72c6-061e-4303-ba35-268d889ca22a.png)
Pyramid layer (smaller) with noise:
![WithNoise](https://user-images.githubusercontent.com/73162709/129896772-226e30b4-10ae-4732-b3d6-10ef337f3004.png)

- Does anyone have any idea where this is coming from?

According to:

    https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=28341

_A combination of the Laplacian and Gaussian is used to make a pyramid processing operation for multi-resolution techniques such as low pass, band pass and high pass combinations._

- Is a Tiled Pyramid TIFF then a [Gaussian pyramid](https://en.wikipedia.org/wiki/Pyramid_(image_processing)) ?

Many thanks already

**Comment by fmw42** [comment]
It does not have to be a Gaussian pyramid. It could simply be an normal average or even decimation. I do not know what IM has coded and it is likely coded in libtiff. But can you post the source image so we can test with it and examine it to see where the "noise" arises? Also post your exact command line for the processing you use for the PTIF output.

**Reply by fmw42** [reply]
Also please identify your version of ImageMagick and platform/OS and version of libtiff.

**Reply by fmw42** [reply]
I processed a 1600 × 1195 JPG image with the same command and it works perfectly fine using IM 6.9.12-19 Q16 Mac OSX with libtiff 4.3.0. I see no noise added and each smaller version is slightly blurred by the pyramid process.

**Reply by ghost** [reply]
First of all, thank you very much for your answer.

We use the following versions:

ImageMagick: ImageMagick 6.9.10-23 Q16 x86_64
OS: Debian 10

I forgot to mention one point. When viewing we filter out certain pixel values.
Our workflow is as follows:
1.) We calculate the Tiled Pyramid TIFF from the raw tiff image (16bit grayscale). 
2.) We calculate the histogram from the raw tiff image.
3.) We filter the pixel values from the raw tiff image based on the histogram. All pixel values outside a certain corridor (too dark or too light values) are set to zero.
4.) We look at the image and zoom in and out respectively we look at different layers and tiles of the pyramid and apply the above filter to the pixel values.

I'm sorry but I can't upload the original image for legal reasons.

What surprises me is that when you zoom out (Change to a higher layer in the pyramid), more values are added and they are brighter than the surroundings.

**Comment by snibgo** [comment]
Without seeing your input file and code, diagnosis is difficult.

If you execute just the ImageMagick command without your filter processing, and view the images in the output tiff with a conventional viewer, does the problem occur? If it does, then we have eliminated your processing as the problem, so the problem is in the input file or ImageMagick. But if the problem isn't visible, then the problem is in your processing.

**Comment by fmw42** [comment]
If you filter different pixels in each resolution, then you will get that kind of issue. The filtering should be in the original and then the pyramid created.

**Reply by fmw42** [reply]
Why are you replacing pixels with black?  Perhaps a better option would be to replace with the median of the 3x3 neighborhood.  Find the "bad" pixels and make a mask using them. Then process the whole resolution with a 3x3 median, then use the mask to merge the original and the median so that only the masked pixels use the median.

**Comment by ghost** [comment]
@snibgo
The problem arises only when we filter the pixel values.

@fmw42 
We will try your suggestion with the replacement.

Thanks for all the feedback. I think we need to take a closer look at the histograms for each layer. I also suspect that we need to adjust the filter for each layer and try the replacement of pixel values suggested by @fmw42.
